compiler/dialects/snax.py

```diff
@@ -15,4 +16,52 @@ class ClusterSyncOp(IRDLOperation):
     name = "snax.cluster_sync_op"
 
 
-Snax = Dialect("snax", [ClusterSyncOp], [])
+@irdl_op_definition
+class ReShuffleOp(IRDLOperation):
+    """ReShuffle operation for memrefs in a snax cluster. This
+    operation is used to change the layout of the memref data"""
+
+    name = "snax.reshuffle"
+
+    source = operand_def(MemRefType[Attribute] | UnrankedMemrefType[Attribute])
+    dest = result_def(MemRefType[Attribute] | UnrankedMemrefType[Attribute])
+
+    def __init__(
+        self,
+        source: SSAValue | Operation,
+        dest: MemRefType[Attribute] | UnrankedMemrefType[Attribute],
+    ):
+        super().__init__(operands=[source], result_types=[dest])
+
+    @staticmethod
+    def from_type_and_target_layout(
+        source: SSAValue | Operation,
+        layout: Attribute,
+    ) -> ReShuffleOp:
+        assert isinstance(source.type, MemRefType)
+        dest = MemRefType(
+            source.type.get_element_type(),
+            shape=source.type.get_shape(),
+            layout=layout,
+            memory_space=source.type.memory_space,
+        )
+        return ReShuffleOp(source, dest)
+
+    def verify_(self) -> None:
+        source = cast(MemRefType[Attribute], self.source.type)
+        dest = cast(MemRefType[Attribute], self.dest.type)
+        if source.get_shape() != dest.get_shape():
+            raise VerifyException(
+                "Expected source and destination to have the same shape."
+            )
+        if source.get_element_type() != dest.get_element_type():
+            raise VerifyException(
+                "Expected source and destination to have the same element type."
+            )
+        if source.memory_space != dest.memory_space:
+            raise VerifyException(
+                "Expected source and destination to have the same memory space."
+            )
```

Comment:
It seems like we can just submit this to upstream with the `memref.cast` operation?

Reply:
Maybe, but can not be ported over as entire block.